rllib/examples/envs/env_rendering_and_recording.py

```diff
@@ -1,135 +1,274 @@
+        # a very small dimension (to save space and bandwidth).
+        image = resize(image, 64, 96)
+        # For WandB videos, we need to put channels first.
+        image = np.transpose(image, axes=[2, 0, 1])
+        # Add the compiled single-step image as temp. data to our Episode object.
+        # Once the episode is done, we'll compile the video from all logged images
+        # and log the video with the EnvRunner's `MetricsLogger.log_...()` APIs.
+        # See below:
+        # `on_episode_end()`: We compile the video and maybe store it).
+        # `on_sample_end()` We log the best and worst video to the `MetricsLogger`.
+        episode.add_temporary_timestep_data("render_images", image)
+
+    def on_episode_end(
+        self,
+        *,
+        episode,
+        env_runner,
+        env,
+        env_index,
+        rl_module,
+        **kwargs,
+    ) -> None:
+        """Computes episode's return and compiles a video, iff best/worst in this iter.
+
+        Note that the actual logging to the EnvRunner's MetricsLogger only happens
+        at the very env of sampling (when we know, which episode was the best and
+        worst). See `on_sample_end` for the implemented logging logic.
         """
+        # Get the episode's return.
+        episode_return = episode.get_return()
+
+        # Better than the best or worse than worst Episode thus far?
+        if (
+            episode_return > self.best_episode_and_return[1]
+            or episode_return < self.worst_episode_and_return[1]
+        ):
+            # Pull all images from the temp. data of the episode.
+            images = episode.get_temporary_timestep_data("render_images")
 
-        # Just generate a random image here for demonstration purposes.
-        # Also see `gym/envs/classic_control/cartpole.py` for
-        # an example on how to use a Viewer object.
-        return np.random.randint(0, 256, size=(300, 400, 3), dtype=np.uint8)
+            # Create a video from the images by simply stacking them.
+            video = np.expand_dims(
+                np.stack(images, axis=0), axis=0
+            )  # TODO: test to make sure WandB properly logs videos.
+            # video = np.stack(images, axis=0)
 
+            # `video` is from the best episode in this cycle (iteration).
+            if episode_return > self.best_episode_and_return[1]:
+                self.best_episode_and_return = (video, episode_return)
+            # `video` is worst in this cycle (iteration).
+            else:
+                self.worst_episode_and_return = (video, episode_return)
+
+    def on_sample_end(
+        self,
+        *,
+        env_runner,
+        metrics_logger,
+        samples,
+        **kwargs,
+    ) -> None:
+        """Logs the best and worst video to this EnvRunner's MetricsLogger."""
+        # Best video.
+        metrics_logger.log_value(
+            "episode_videos_best",
+            self.best_episode_and_return[0],
```

Comment:
More sense to me makes to log the best of the best and the worst of the horst.

Reply:
Images alone are already large, videos are larger and then multiple of them ... my machine would probably not make it well.

Reply:
Yeah, that's why I crunch them down to 64 x 86. True, we should maybe further reduce them on the Algorithm side after the EnvRunner each return their best, but that would require another callback that goes through these, etc..
I wanted to avoid this complexity. One could also simply log only the videos from one hard-coded EnvRunner (e.g. index=1) and reduce the number of videos this way. Or only log every nth iteration, etc..